Added docs for token reset (#273)
AxonIQ/reference-guide: axon-server/administration/reset-event-processor-token.md

> # How to reset the Token of an Event Processor

Reviewer: Would it be feasible to rename this page to `event-processors.md`, so that we open up the file for other Event Processor administration sections later on? Just a section/page on resetting out of the blue feels slightly off to me  perhaps :thinking:

It might be desirable to reset the position an event processor works on.
This is done by resetting its token.

We document the following ways to reset the token of an event processor:

> - Using Axon Framework

Reviewer: Note the reset invocation from Axon Framework is already described [here](https://docs.axoniq.io/reference-guide/axon-framework/events/event-processors/streaming#triggering-a-reset). It's not wrong to rehash it here, but I would definitely reference the already existing section.

This requires all instances of the event processor (on all nodes) to release their claims on segments.
Now, when `resetTokens()` is called, the executing instance can temporarily claim all segments, reset the tokens and store the new values.
After the token has been reset, the event processor can be started again.

Note that older versions of axon framework (pre 4.6) directly acknowledge a received admin instruction, without waiting for its execution.
This might require manually waiting for an EventProcessor to be fully stopped before the token can be reset.
A possible approach for this is shown in the last section of this page.

## Using Axon Framework
Axon Framework exposes all required functionality for a reset in the `StreamingEventProcessor` class. In the following, we provide a sample for its usage. For more details, read the documentation on the framework classes [here](../../axon-framework/events/event-processors/streaming.md#triggering-a-reset).
Instances of `StreamingEventProcessor` can be obtained by querying the `eventProcessingConfiguration` method of your global configuration.
```java

> configuration.eventProcessingConfiguration()

Reviewer: I think it would be good to define what type the `configuration` is here. 
Right now, users may copy this section and don't know where to find the `configuration` to use here.

In training material, I'd even add the import statement for clarity, although that might not be necessary here.

            streamingEventProcessor.resetTokens();
            streamingEventProcessor.start();
        }
    });
```

Note that this only concerns stopping the local instances. 
If there are instances running on other nodes, you either need to use AxonServer or build a solution for this yourself.
For more details, refer to the section in the Axon Framework [reference guide](../../axon-framework/events/event-processors/streaming.md#triggering-a-reset).

## Using the REST API
Axon Server exposes a REST API to pause and start EventProcessors.
When used in conjunction with the `resetTokens` method from earlier, this can be used to make sure that all instances of an EventProcessor are paused before a token is reset.
The required parameters to do this are the following:
- *component*: name of the component the EventProcessor is part of
- *processor*: name of the EventProcessor itself
- *context*: name of the context for which to reset the tokens
- *tokenStoreId*: the identifier used to distinguish the desired token from other tokens stored in the same store. 

Now 
`/v1/components/{component}/processors/{processor}/pause?context={context}&tokenStoreIdentifier={tokenStoreId}`

> can be called, causing AxonServer to request all matched EventProcessors to stop.

Reviewer: Similar "space or no space" argument as with Event Processors. Furthermore, you're using "Axon Server" at the start of this section. 

Personally, I opt for using a space between Axon and Server.
But, consistency is more important than my personal preference here if you ask me.

> Then a reset as shown in the example using the framework can be performed.

Reviewer: Nit: would be nice to use a reference to that previous section here. Maybe users jumped over to the REST approach immediately. Hence, they cannot reminiscence over the earlier explained Axon Framework solution at all. By providing a link we simplify that step for those readers.

This is not Axon Server specific and hence is the same for all shown methods.

As a final step, the EventProcessors can be started again with a patch request to the following url:
`/v1/components/{component}/processors/{processor}/start?context={context}&tokenStoreIdentifier={tokenStoreId}` .

## Using the Axon Server Connector
To reset EventProcessors with the Axon Server Connector, the dependency has to be available on your classpath, 
e.g. by using the following maven dependency.
```xml
<dependency>
    <groupId>io.axoniq</groupId>
    <artifactId>axonserver-connector-java</artifactId>

>     <version>4.6.0-SNAPSHOT</version>

Reviewer: As this commit points to `master`, this will become the 4.6 documentation page. That means you can be certain there's a `4.6.0` release.
```suggestion
    <version>4.6.0</version>
```

</dependency>
```

All operations related to administration go through an `AdminChannel`, 
which can be obtained from an `AxonServerConnectionFactory` as follows.

```java
private AdminChannel adminChannel() {
    AxonServerConnectionFactory connectionFactory = AxonServerConnectionFactory.forClient(componentName).build();
    return connectionFactory.connect(contextName).adminChannel();
}
```

In this example, `componentName` and `contextName` are values supplied by external configuration.

> In simple cases, these might be the same as in your axon-framework configuration.

Reviewer: ```suggestion
In simple cases, these might be the same as in your Axon Framework configuration.
```

Using the provided admin channel, you can pause, reset and restart the event processors.

> In contrast to the approach based on the Axon Framework, you also need to provide a `tokenStoreIdentifier`, 

Reviewer: Might be nice to tell users how they can retrieve the `tokenStoreIdentifier`. Might be good even to add this notice on the "Streaming Event Processors" page's Token Store section, as it's currently not drafted up there at all. Having a piece on it there gives you the opportunity to link to it from this page.

since there can be multiple applications connected to one Axon Server that share the same token store. 

Now resetting the tokens can be done by simply calling the `pauseEventProcessor`, `resetTokens` and `startEventProcessor` 
methods in the correct order.

```java
adminChannel().pauseEventProcessor(eventProcessorName, tokenStoreIdentifier)
        .thenRun(eventProcessor::resetTokens)
        .thenRun(() -> adminChannel().startEventProcessor(eventProcessorName, tokenStoreIdentifier))
```

Note that Axon Server makes sure to stop and start all matched EventProcessors on all nodes with this call. 

## Handling asynchronous behaviour in older Axon Framework versions (pre 4.6)

Prior to version 4.6, Axon Framework did immediately acknowledge receiving a pause instruction.
This means, that the Axon Server would receive these `ACCEPTED` Results before all EventProcessors have terminated.

> Starting from Axon Framework version 4.6, the Axon Server will only respond with `SUCCESS`, once all connected EventProcessors have successfully been paused.

Reviewer: ```suggestion
Starting from Axon Framework version 4.6, Axon Server will only respond with `SUCCESS` once all connected Event Processors have successfully been paused.
```
